pkg/controller/dns/resource_record_set.go

```diff
@@ -0,0 +1,252 @@
+	projectID, opts, err := gcp.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, err
+	}
+	d, err := dns.NewService(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	return &external{
+		kube:      c.kube,
+		dns:       d.ResourceRecordSets,
+		projectID: projectID,
+	}, nil
+}
+
+type external struct {
+	kube      client.Client
+	dns       *dns.ResourceRecordSetsService
+	projectID string
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotResourceRecordSet)
+	}
+
+	rrs, err := e.Get(ctx, cr)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(
+			resource.Ignore(gcp.IsErrorNotFound, err),
+			errGetFailed,
+		)
+	}
+
+	lateInit := false
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	rrsClient.LateInitializeSpec(&cr.Spec.ForProvider, *rrs)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errManagedUpdateFailed)
+		}
+		lateInit = true
+	}
+
+	cr.Status.AtProvider = rrsClient.GenerateObservation(*rrs)
+	cr.SetConditions(xpv1.Available())
+
+	upToDate, err := rrsClient.IsUpToDate(meta.GetExternalName(cr), &cr.Spec.ForProvider, rrs)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errCheckUpToDate)
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        upToDate,
+		ResourceLateInitialized: lateInit,
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotResourceRecordSet)
+	}
+
+	args := &dns.ResourceRecordSet{}
+	rrsClient.GenerateResourceRecordSet(
+		meta.GetExternalName(cr),
+		cr.Spec.ForProvider,
+		args,
+	)
+
+	rrs, err := e.dns.Create(
+		e.projectID,
+		cr.Spec.ForProvider.ManagedZone,
+		args,
+	).Context(ctx).Do()
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCannotCreate)
+	}
+
+	cr.Status.AtProvider = rrsClient.GenerateObservation(*rrs)
+	cr.SetConditions(xpv1.Creating())
+
+	return managed.ExternalCreation{}, nil
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.ResourceRecordSet)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotResourceRecordSet)
+	}
+
+	args := &dns.ResourceRecordSet{}
+	rrsClient.GenerateResourceRecordSet(meta.GetExternalName(cr), cr.Spec.ForProvider, args)
+
+	rrs, err := e.dns.Patch(
+		e.projectID,
+		cr.Spec.ForProvider.ManagedZone,
+		cr.Status.AtProvider.Name,
+		cr.Status.AtProvider.Type,
```

Comment:
We should not rely on `Status` for the desired state. I believe we can already get name from external name annotation and type available in `Spec`.
See some related discussion on why relying on status is not something we should avoid: https://github.com/vmware-tanzu/velero/issues/1687

Reply:
I understand that rely on status is a problem @turkenh, but I couldn't think in another solution. 

Here is the case: when the users wants to update `Name` and/or `Type`, in the first reconcile loop it will try to `GET` the RRS using the new `Name` and `Type` values, but it doesn't exist yet (because the external API is with the "old" values), so the `Observe` method will return that the resource doesn't exists. But the desired behavior is: find the RRS with the old values, compare it in the `IsUpToDate` and then update it.

If I use `Spec` fields in the `Update`, it won't update because the resource won't be found

Reply:
You mean updating the name of the external resource, e.g. external-name annotation right? 
IIUC, I am trying to make an analogy like you have a `GKECluster` named as `my-cluster` and want to rename it as `my-old-cluster` by updating external-name annotation.

I believe, this is not a valid use case and Crossplane does not guarantee to keep working in such a case.
Regarding `Type`, I believe we can mark it as [immutable](https://github.com/crossplane/crossplane/blob/d04a059fe341a941760a3a94ef07085ae7158365/design/one-pager-managed-resource-api-design.md#immutable-properties).

Reply:
Yes, your analogy is correct! The problem is when the user wants to update a field that is used as a identifier to the resource. In RRS case, the SDK uses `Name` and `Type` to find which RRS you want to get.

So what you are saying is that crossplane doesn't support update for external names? I thought other managed resources had this logic implemented 😕 
`Type` isn't really a immutable field, you can update it through PATCH/PUT requests... I don't know, feels like both solutions (relying on `Status` or just "not supporting" this use case) are wrong 🤔 

I will remove them from the controller implementation, but don't you think that this use case is pretty common for companies? IMHO renaming resources is a important use case to be covered

Reply:
I would really try to discuss it further in another issue to try to design a workaround to support those use cases, is it a commonly asked feature from customers? Is there some discussion already being made for this?